moodle/Sniffs/Commenting/PHPDocTypesSniff.php

```diff
@@ -0,0 +1,960 @@
+<?php
+
+// This file is part of Moodle - https://moodle.org/
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <https://www.gnu.org/licenses/>.
+
+/**
+ * Check PHPDoc Types.
+ *
+ * @copyright  2024 Otago Polytechnic
+ * @author     James Calder
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later (or CC BY-SA v4 or later)
+ */
+
+namespace MoodleHQ\MoodleCS\moodle\Sniffs\Commenting;
+
+use PHP_CodeSniffer\Sniffs\Sniff;
+use PHP_CodeSniffer\Files\File;
+use MoodleHQ\MoodleCS\moodle\Util\PHPDocTypeParser;
+
+/**
+ * Check PHPDoc Types.
+ */
+class PHPDocTypesSniff implements Sniff
+{
+    /** @var ?File the current file */
+    protected ?File $file = null;
+
+    /** @var array{'code': ?array-key, 'content': string, 'scope_opener'?: int, 'scope_closer'?: int}[]
+     * file tokens */
+    protected array $tokens = [];
```

Comment:
OK, hold off on checking more.  I'll have a go at rewriting the sniff.

Reply:
I've started rewriting to pass the scopes between methods in a meaningful way (and I think it'll be nicer this way).  The rest are really global variables, though, I think.  If it's a technical issue, I could wrap them in an object, and pass the whole lot everywhere.  If it's a style issue, though, then this wouldn't really solve it, because it'd still be globals in spirit.

Reply:
I've rewritten to pass the scopes, and one of the comment variables.  It also no longer stores state between calls to process().